services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedDeviceBackend.java

```diff
@@ -0,0 +1,270 @@
+                });
+    }
+
+    @Override
+    public Future<OperationResult<Id>> updateDevice(final String tenantId, final String deviceId, final Device device,
+            final Optional<String> resourceVersion, final Span span) {
+        return registrationService.updateDevice(tenantId, deviceId, device, resourceVersion, span);
+    }
+
+    // CREDENTIALS
+
+    @Override
+    public final Future<CredentialsResult<JsonObject>> get(final String tenantId, final String type,
+            final String authId) {
+        return credentialsService.get(tenantId, type, authId);
+    }
+
+    @Override
+    public Future<CredentialsResult<JsonObject>> get(final String tenantId, final String type, final String authId,
+            final Span span) {
+        return credentialsService.get(tenantId, type, authId, span);
+    }
+
+    @Override
+    public final Future<CredentialsResult<JsonObject>> get(final String tenantId, final String type,
+            final String authId, final JsonObject clientContext) {
+        return get(tenantId, type, authId, clientContext, NoopSpan.INSTANCE);
+    }
+
+    @Override
+    public  Future<CredentialsResult<JsonObject>> get(final String tenantId, final String type, final String authId, final JsonObject clientContext,
+            final Span span) {
+        return credentialsService.get(tenantId, type, authId, clientContext, span)
+                .compose(result -> {
+                    if (result.getStatus() == HttpURLConnection.HTTP_NOT_FOUND
+                            && isAutoProvisioningEnabled(type, clientContext)) {
+                        return provisionDevice(tenantId, authId, clientContext, span);
+                    }
+                    return Future.succeededFuture(result);
+                });
+    }
+
+    /**
+     * Parses certificate, provisions device and returns the new credentials.
+     */
+    private Future<CredentialsResult<JsonObject>> provisionDevice(final String tenantId, final String authId,
+            final JsonObject clientContext,
+            final Span span) {
+
+        final X509Certificate cert;
+        try {
+            final byte[] bytes = clientContext.getBinary(CredentialsConstants.FIELD_CLIENT_CERT);
+            final CertificateFactory factory = CertificateFactory.getInstance("X.509");
+            cert = (X509Certificate) factory.generateCertificate(new ByteArrayInputStream(bytes));
+
+            if (!cert.getSubjectX500Principal().getName(X500Principal.RFC2253).equals(authId)) {
+                throw new IllegalArgumentException("Subject DN of the client certificate does not match authId");
+            }
+        } catch (final CertificateException | ClassCastException | IllegalArgumentException e) {
+            TracingHelper.logError(span, e);
+            final int status = HttpURLConnection.HTTP_BAD_REQUEST;
+            return Future.succeededFuture(createErrorCredentialsResult(status, e.getMessage()));
+        }
```

Comment:
this seems to be the same code as in the file bsaed registry, isn't it?

Reply:
Yes. My plan is to pull up commonly used properties and such reusable codes to the base module.

Reply:
@sophokles73 I could imagine pulling up this method to the `AutoProvisioningEnabledDeviceBackend` as given below. WDYT?

```
    /**
     * Registers a device together with a set of credentials for the given client certificate.
     *
     * @param tenantId The tenant to which the device belongs.
     * @param authId The authentication identifier.
     * @param clientContext The client context that can be used to get the X.509 certificate 
     *                      of the device to be provisioned.
     * @param span The active OpenTracing span for this operation. It is not to be closed in this method! An
     *            implementation should log (error) events on this span and it may set tags and use this span as the
     *            parent for any spans created in this method.
     * @return A (succeeded) future containing the result of the operation. The <em>status</em> will be
     *         <ul>
     *         <li><em>201 CREATED</em> if the device has successfully been provisioned.</li>
     *         <li><em>4XX</em> if the provisioning failed. The payload may contain an error description.</li>
     *         </ul>
     * @throws NullPointerException if any of the parameters is {@code null}.
     */
    default Future<OperationResult<String>> provisionDevice(
            final String tenantId,
            final String authId,
            final JsonObject clientContext,
            final Span span) {
        // TODO.
        return null;

    }
```

Reply:
I was more thinking about pulling up a utility method for parsing the client context ...

Reply:
I have now pulled up a utility method for getting the certificate from the client context. Once this is merged, I will raise a separate PR for refactoring the file based device registry with this new utility method.